// directus/dist/services/roles.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RolesService = void 0;
const items_1 = require("./items");
const permissions_1 = require("./permissions");
const users_1 = require("./users");
const presets_1 = require("./presets");
const exceptions_1 = require("../exceptions");
class RolesService extends items_1.ItemsService {
    constructor(options) {
        super('directus_roles', options);
    }
    deleteOne(key) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.deleteMany([key]);
            return key;
        });
    }
    deleteMany(keys) {
        const _super = Object.create(null, {
            deleteMany: { get: () => super.deleteMany }
        });
        return __awaiter(this, void 0, void 0, function* () {
            // Make sure there's at least one admin role left after this deletion is done
            const otherAdminRoles = yield this.knex
                .count('*', { as: 'count' })
                .from('directus_roles')
                .whereNotIn('id', keys)
                .andWhere({ admin_access: true })
                .first();
            const otherAdminRolesCount = +((otherAdminRoles === null || otherAdminRoles === void 0 ? void 0 : otherAdminRoles.count) || 0);
            if (otherAdminRolesCount === 0)
                throw new exceptions_1.UnprocessableEntityException(`You can't delete the last admin role.`);
            yield this.knex.transaction((trx) => __awaiter(this, void 0, void 0, function* () {
                const permissionsService = new permissions_1.PermissionsService({
                    knex: trx,
                    accountability: this.accountability,
                    schema: this.schema,
                });
                const presetsService = new presets_1.PresetsService({
                    knex: trx,
                    accountability: this.accountability,
                    schema: this.schema,
                });
                const usersService = new users_1.UsersService({
                    knex: trx,
                    accountability: this.accountability,
                    schema: this.schema,
                });
                // Delete permissions/presets for this role, suspend all remaining users in role
                yield permissionsService.deleteByQuery({
                    filter: { role: { _in: keys } },
                });
                yield presetsService.deleteByQuery({
                    filter: { role: { _in: keys } },
                });
                yield usersService.updateByQuery({
                    filter: { role: { _in: keys } },
                }, {
                    status: 'suspended',
                    role: null,
                });
                yield _super.deleteMany.call(this, keys);
            }));
            return keys;
        });
    }
    delete(key) {
        return __awaiter(this, void 0, void 0, function* () {
            if (Array.isArray(key))
                return yield this.deleteMany(key);
            return yield this.deleteOne(key);
        });
    }
}
exports.RolesService = RolesService;
