// directus/dist/database/migrations/20201105A-add-cascade-system-relations.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.down = exports.up = void 0;
const updates = [
    {
        table: 'directus_fields',
        constraints: [
            {
                column: 'group',
                references: 'directus_fields.id',
            },
        ],
    },
    {
        table: 'directus_files',
        constraints: [
            {
                column: 'folder',
                references: 'directus_folders.id',
            },
            {
                column: 'uploaded_by',
                references: 'directus_users.id',
            },
            {
                column: 'modified_by',
                references: 'directus_users.id',
            },
        ],
    },
    {
        table: 'directus_folders',
        constraints: [
            {
                column: 'parent',
                references: 'directus_folders.id',
            },
        ],
    },
    {
        table: 'directus_permissions',
        constraints: [
            {
                column: 'role',
                references: 'directus_roles.id',
            },
        ],
    },
    {
        table: 'directus_presets',
        constraints: [
            {
                column: 'user',
                references: 'directus_users.id',
            },
            {
                column: 'role',
                references: 'directus_roles.id',
            },
        ],
    },
    {
        table: 'directus_revisions',
        constraints: [
            {
                column: 'activity',
                references: 'directus_activity.id',
            },
            {
                column: 'parent',
                references: 'directus_revisions.id',
            },
        ],
    },
    {
        table: 'directus_sessions',
        constraints: [
            {
                column: 'user',
                references: 'directus_users.id',
            },
        ],
    },
    {
        table: 'directus_settings',
        constraints: [
            {
                column: 'project_logo',
                references: 'directus_files.id',
            },
            {
                column: 'public_foreground',
                references: 'directus_files.id',
            },
            {
                column: 'public_background',
                references: 'directus_files.id',
            },
        ],
    },
    {
        table: 'directus_users',
        constraints: [
            {
                column: 'role',
                references: 'directus_roles.id',
            },
        ],
    },
];
/**
 * NOTE:
 * Not all databases allow (or support) recursive onUpdate/onDelete triggers. MS SQL / Oracle flat out deny creating them,
 * Postgres behaves erratic on those triggers, not sure if MySQL / Maria plays nice either.
 */
function up(knex) {
    return __awaiter(this, void 0, void 0, function* () {
        for (const update of updates) {
            yield knex.schema.alterTable(update.table, (table) => {
                for (const constraint of update.constraints) {
                    table.dropForeign([constraint.column]);
                    table.foreign(constraint.column).references(constraint.references);
                }
            });
        }
    });
}
exports.up = up;
function down(knex) {
    return __awaiter(this, void 0, void 0, function* () {
        for (const update of updates) {
            yield knex.schema.alterTable(update.table, (table) => {
                for (const constraint of update.constraints) {
                    table.dropForeign([constraint.column]);
                }
            });
        }
    });
}
exports.down = down;
